### PR TITLE
net: don't lock cs_main while reading blocks in net processing

### PR DESCRIPTION
**Pull request by andrewtoth**
Inspired by https://github.com/bitcoin/bitcoin/pull/11913 and https://github.com/bitcoin/bitcoin/pull/26308.

`cs_main` doesn't need to be locked while reading blocks. This removes the locks in `net_processing`.

### CODE REVIEW

**Comment by dergoegge** [comment]
Concept ACK

I would suggest that you rebase this PR on top of #26316 (if there is a dependency) and then maybe also mark this PR as a draft until the base is merged.

File: src/net_processing.cpp
Comment (MarcoFalke) [inline src/net_processing.cpp:3872]: I don't think this can happen. MAX_BLOCKTXN_DEPTH is 10, so could keep the assert or at least make it a error log category +  Assume(false)?

File: src/node/blockstorage.h
Comment (MarcoFalke) [inline src/node/blockstorage.h:214]:
```suggestion
bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const CBlockIndex& block, const CMessageHeader::MessageStartChars& message_start);
```

shouldn't the `*` be a `&`, like before? (`nullptr` is not accepted)

File: src/net_processing.cpp
Comment (andrewtoth) [reply src/net_processing.cpp:3872]: Could it not happen in regtest?

Comment (MarcoFalke) [reply src/net_processing.cpp:3872]: Not sure, but it seems fine to crash in tests if they violate assumptions

Comment (MarcoFalke) [reply src/net_processing.cpp:3872]: See `MIN_BLOCKS_TO_KEEP`, which is also enforced in regtest

**Comment by jamesob** [comment]
Cool. Can someone add a "needs benchmark" label?

**Comment by andrewtoth** [comment]
> Can someone add a "needs benchmark" label?

What kind of benchmark would be appropriate here? AFAICT message processing is done on a single thread, so this patch wouldn't speed up handling requests. It would free up other threads that are waiting on the lock as the message processing thread does lengthy disk IO.

**Review by hernanmarino — COMMENTED** [review]
Approach ACK. Run all tests, everything run fine.

**Review by pablomartin4btc — COMMENTED** [review]
cr ACK. I like the idea and I think this change is very useful for performance purposes. 
I ran unit tests and functional ones with no errors or warnings.
Pending of benchmarking if I manage to find the way, otherwise when someone posts here I'm happy to re-evaluate.

**Comment by andrewtoth** [comment]
I wrote a tool to allow benchmarking the affected code paths https://github.com/andrewtoth/spam_block_reqs. I ran it on block 768022 using 5000 requests for each of the four paths on both this branch and master. For example:
```
cargo run --release -- -r block-transactions -b 000000000000000000064b1c6e9606714fd4a96d6ff877e4a93b170d86361e28 -n 5000
```

I did not benchmark the `MSG_FILTERED_BLOCK` path because it isn't supported in `rust-bitcoin`, however there isn't anything in that path that requires taking the lock other than the shared code for the other paths.

Results show no regression:
| Request Type | master (cb32328d1b80d0ccd6eb9532bd8fe4e0a4de385e) | branch |
| -------------------- |-------------------------------------------------------------------------------------|------------|
| witness-block | 22.6s | 22.3s |
| compact-block | 60.9s | 59.1s |
| block-transactions | 58.5s | 59.0s |
| legacy-block | 77.7s | 77.4s |

However, while running the tool using a very high number of requests so it would not complete, I ran a benchmark requesting 2000 headers from block 750k 1000 times on a single core using REST. For all the request-types in this branch the mean response was 1ms, the same as idle. While running the tool on master, the mean response times were:
witness-block - 14ms
compact-block - 19ms
block-transactions - 19ms
loegacy-block - 42ms

So it seems like this definitely improves responsiveness for any other threads that need to acquire a lock to `cs_main`.

File: src/node/blockstorage.h
Comment (andrewtoth) [reply src/node/blockstorage.h]: Removed touching this file.

File: src/net_processing.cpp
Comment (andrewtoth) [reply src/net_processing.cpp]: Updated this to keep the assert. Also changed it to take `FlatFilePos` in `ReadBlockFromDisk` so it doesn't have to reacquire the lock again after releasing it.

**Comment by andrewtoth** [comment]
Moved all whitespace only scope changes to separate commit to make it easier to review.

File: src/net_processing.cpp
Comment (furszy) [inline src/net_processing.cpp:3877]: What about this assertion? Couldn't `ReadBlockFromDisk` return false if the block is not on disk anymore?

Comment (andrewtoth) [reply src/net_processing.cpp:3877]: See https://github.com/bitcoin/bitcoin/pull/26326#discussion_r1043171470. Blocks below `MAX_BLOCKTXN_DEPTH` can't be pruned, so we should retain the current behavior if this read fails.

File: src/net_processing.cpp
Comment (furszy) [inline src/net_processing.cpp]: tiny nit: would be good to add a jump line here.

File: src/net_processing.cpp
Comment (furszy) [inline src/net_processing.cpp:2177]: tiny nit: would be good to initialize the variable to nullptr and false.

**Review by furszy — COMMENTED** [review]
Code ACK 3e93e6b7

**Review by furszy — COMMENTED** [review]
re-ACK 144dd46.
No changes since the last review.

**Comment by andrewtoth** [comment]
@furszy thank you for the review. Sorry I forgot to address your nits on this rebase. Will do them if I have to rebase again.

**Review by furszy — COMMENTED** [review]
reACK e8357eb after rebase.

**Review by furszy — COMMENTED** [review]
diff-ACK fae1fd61

Have to admit that I'm not fan of the first commit. It forces me to remember where the indentation changes were located when I'm reviewing the other commits.

File: src/net_processing.cpp
Comment (furszy) [reply src/net_processing.cpp:2315]: > Last commit: Not sure. This looks like a p2p protocol change?

For the local node, this change means that the node will continue functional instead of crashing. 
Which I think is good. The node shouldn't crash because it received a p2p message that cannot answer.

For the remote peer, this means that the connection will remain active for ~10 more minutes until the block request times out and the peer closes the socket.
Which isn't the best, but it is how we currently handle unanswered block requests.

Are you seeing something else?

Comment (furszy) [reply src/net_processing.cpp:2315]: > > For the local node, this change means that the node will continue functional instead of crashing.
> 
> The node wouldn't crash here, because cs_main was being held, so no pruning could happen in-between.

Could also be that the requested block data is not on disk or not accessible for the time this happen. Something external to our software.

Still, removing the external cause from the equation. The only diff here is that we allow a tiny extra pruning window in-between the index lookup and the block data read.
Because, we currently return early if the block is pruned [here](https://github.com/bitcoin/bitcoin/blob/e789b30b2565c8bdbf48a45f2c5a7b92e5d61d25/src/net_processing.cpp#L2285), without logging anything. And after this PR, we could either return early on the no logging line (if the block is already pruned) or log "Cannot load block from disk. It was likely pruned before we could read it" if the block was pruned right after the index lookup.

Other than that, I'm not seeing any p2p protocol change.

Maybe @andrewtoth could take this to the PR description?

Comment (furszy) [reply src/net_processing.cpp:2315]: > If your datadir was corrupted from outside, I think a crash is acceptable. You likely wouldn't be able to re-org or otherwise stay in consensus anyway.

You know; if the block in conflict is old enough, then the node should be "mostly" ok. The problem would be on new peers performing IBD, which would disconnect from the node for the lack of response. 

> > I'm not seeing any p2p protocol change.
> 
> Maybe "protocol change" was the wrong word? I guess "behavior change" may be better?

Sounds good.

Comment (andrewtoth) [reply src/net_processing.cpp:2315]: If a block is pruned before we lock cs_main, this function returns early. This behavior is not changed in this PR.

If a block is not pruned before we lock cs_main, previously we are guaranteed to send the block to the peer.

This PR changes this to have a short window after unlocking cs_main but before opening the fd to read the block that if the block is pruned and unlinked from another thread, then this function also returns early.

I think we all agree this is not a p2p protocol change? I also believe I documented this behavior change in the PR description and in the debug logs.

Comment (andrewtoth) [reply src/net_processing.cpp:2315]: So the other behavior change is that if we fail to read a block for reasons other than being pruned, previously the node would crash. Now we will log the failure and continue.

> If your datadir was corrupted from outside, I think a crash is acceptable. You likely wouldn't be able to re-org or otherwise stay in consensus anyway.

I suppose if the read fails, we can retake cs_main and check if the block was pruned out from under us. If so, we log and continue, otherwise we can keep crashing on an assert? Do we really want to keep crashing here if we fail to read a block? I'm not sure.

Comment (andrewtoth) [reply src/net_processing.cpp]: > if the read fails, we can retake cs_main and check if the block was pruned out from under us. If so, we log and continue, otherwise we can keep crashing on an assert

@maflcko I updated the code to do this. Thus I don't think there is any behavior change.
If the block requested is pruned, we return early before and after this PR. If the block requested is unpruned and we fail to read it, we will crash before and after this PR.

Comment (furszy) [reply src/net_processing.cpp]: >  I am not asking to add back the assert.

+1 on not re-adding the assert.

> However, I wonder what the point is of forcing the remote peer into a 10 minute delay+disconnect. Why not disconnect immediately? Also, if the remote has set a permission flag to avoid the disconnect (not sure if this exists), I wonder what the correct behavior would be.

If the remote peer has a permission flag to avoid the disconnection, maybe it could respond with a `NOTFOUND` message? Same as we do with tx getdata requests.
But maybe better to discuss this in a follow-up or in an issue?

Comment (andrewtoth) [reply src/net_processing.cpp:2323]: Updated to log and return instead of hitting an assert. Also updated the commit description to clarify this behavior.

> However, I wonder what the point is of forcing the remote peer into a 10 minute delay+disconnect. Why not disconnect immediately?

We could, but I think this is very unlikely to occur in practice. We already disconnect above on line 2293 if we are past the NODE_NETWORK_LIMITED_MIN_BLOCKS threshold, which is MIN_BLOCKS_TO_KEEP (288) so we won't prune below that anyways.

**Review by furszy — COMMENTED** [review]
rebase utACK 1027339f

File: src/net_processing.cpp
Comment (darosior) [inline src/net_processing.cpp:4077]: Worth adding a comment here, it's not immediate that this assertion would never fail because `MAX_BLOCKTXN_DEPTH < MIN_BLOCKS_TO_KEEP`.

Comment (andrewtoth) [reply src/net_processing.cpp]: Rebased and added a comment to address this.

**Review by furszy — COMMENTED** [review]
Re-ACK f9d4152.
Since my last review, this PR was only rebased and a comment was added.

File: src/net_processing.cpp
Comment (sr-gi) [reply src/net_processing.cpp]: Would it be woth having a static assert for this when defining `MAX_BLOCKTXN_DEPTH`?

```
static_assert(MAX_BLOCKTXN_DEPTH < MIN_BLOCKS_TO_KEEP, "MIN_BLOCKS_TO_KEEP too low");
```

This was not obvious to me either, but I'm not too familiar with this part of net_processing

Comment (sr-gi) [reply src/net_processing.cpp]: Would it be worth having a static assert for this when defining `MAX_BLOCKTXN_DEPTH`?

```
static_assert(MAX_BLOCKTXN_DEPTH < MIN_BLOCKS_TO_KEEP, "MIN_BLOCKS_TO_KEEP too low");
```

This was not obvious to me either, but I'm not too familiar with this part of net_processing

**Review by sr-gi — COMMENTED** [review]
Concept ACK

I agree with @furszy in https://github.com/bitcoin/bitcoin/pull/26326#discussion_r1381776328 that we should followup this with sending NotFounds for all the blocks that we are unable (or unwilling) to provide

**Comment by andrewtoth** [comment]
@sr-gi thank you for your review!

> that we should followup this with sending NotFounds for all the blocks that we are unable (or unwilling) to provide

After https://github.com/bitcoin/bitcoin/pull/28120 I don't think it is very likely for this to occur by updated nodes. For older nodes and other software that requests blocks past the prune threshold it would require a protocol change as @sipa mentioned.

**Comment by furszy** [comment]
Little add about the notfound response; I already have a draft for it. I'm busy with other stuff this week but will open the PR next week :).

File: src/net_processing.cpp
Comment (mzumsande) [inline src/net_processing.cpp:2313]: As written somewhere above, we never prune blocks > 288 blocks from the tip, and if pruning, we should have disconnected any peer requesting older blocks - so it should be impossible to trigger this log even with pruning.
On the other hand, it could still be hit while not pruning due to disk problems - e.g.  helping other peers with IBD. Searching for this assert gives me multiple matches, e.g.
* #6263
* #5668
* #12230
* https://bitcoin.stackexchange.com/questions/113218/how-to-fix-the-bitcoind-service-shutdown-seemingly-by-invalid-block

It apparently also crashed forks like Bitcoin Unlimited back in the days.

I'm not against removing it, but I think that if we do, we should at least have an error message that mentions the possibility of disk failures, and that is also unconditional so it doesn't get hidden in the spammy NET log that very few people enable.

**Review by mzumsande — COMMENTED** [review]
Concept ACK

File: src/net_processing.cpp
Comment (andrewtoth) [reply src/net_processing.cpp:2313]: > so it should be impossible to trigger this log even with pruning

Since we have a 2 block buffer, peers can request up to 290 blocks. So one way this can be triggered - a peer requests block 290 and we get the filepos, we release the lock, `pruneblockchain` RPC is called, rpc thread acquires the lock and block 290 is pruned before we can acquire the fd, read now fails.
However, `pruneblockchain` has a 2 hour buffer, so there must also be a long window of no new blocks.
I don't think this is really a feasible scenario though.

Comment (mzumsande) [reply src/net_processing.cpp:2313]: In that case, maybe have a different and unconditional error message / keep the assert just for the case where pruning isn't enabled?

Comment (andrewtoth) [reply src/net_processing.cpp:2313]: Indeed, I think that's the best way. I did that before https://github.com/bitcoin/bitcoin/pull/26326#discussion_r1375494479 but other reviewers thought removing the assert is better. But, I think it makes it easier to merge this PR if there is strictly no behavior change, and other patches can be proposed to modify the assertion behavior or disconnecting instead of returning early.

So, if reading the block fails, we retake `cs_main` and determine if the block was pruned out from under us. If so, we log and return early. Otherwise, we assert as before.

@maflcko @furszy what do you think?

Comment (mzumsande) [reply src/net_processing.cpp:2313]: > determine if the block was pruned out from under us

That sounds like work, not sure how easy it is - I just meant to check whether we're running in pruning mode, so the block could have been pruned theoretically. I don't have a strong opinion between asserting / logging an error unconditionally, I just think that a conditional NET log is not sufficient in a case where something is seriously wrong on our end.

Comment (furszy) [reply src/net_processing.cpp:2313]: > I just meant to check whether we're running in pruning mode, so the block could have been pruned theoretically.

Just in case, it should also check that the block is further than the pruning threshold.

> So, if reading the block fails, we retake cs_main and determine if the block was pruned out from under us. If so, we log and return early. Otherwise, we assert as before.

For now (at least in this PR), we should maintain the current network behavior if the assert is replaced. This means disconnecting the node immediately when it happens (same behavior as if the node crashed. The socket gets closed).

**Comment by andrewtoth** [comment]
Rebased and updated
- Initial whitespace only commit has been removed
- Added `static_assert(MAX_BLOCKTXN_DEPTH <= MIN_BLOCKS_TO_KEEP, "MAX_BLOCKTXN_DEPTH too high");` in first commit
- In the second commit, if reading the block fails, we retake `cs_main` and check if the block has been pruned. If so, we conditionally log to `NET` category, otherwise we unconditionally log an error. In either case, we also disconnect the peer before returning.

Thank you for your reviews @furszy @sr-gi @mzumsande !

**Comment by sr-gi** [comment]
ACK [75d27fe](https://github.com/bitcoin/bitcoin/pull/26326/commits/75d27fefc7a04ebdda7be5724a014b6a896e7325)

The only differences are the ones mentioned, which align with the approach after considering https://github.com/bitcoin/bitcoin/pull/26326#discussion_r1588242121

File: src/net_processing.cpp
Comment (furszy) [inline src/net_processing.cpp:2479]: nit: It wouldn't hurt to add a comment above this disconnection (just mention how this behaved historically).
I think that people reading this code for the first time will not understand why the peer is being disconnected for a node local failure.

**Review by furszy — COMMENTED** [review]
ACK 75d27fefc with a non-blocking nit.

File: src/net_processing.cpp
Comment (andrewtoth) [reply src/net_processing.cpp:2479]: There is a comment a few lines up that describes the behavior https://github.com/bitcoin/bitcoin/pull/26326/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R2453. But I agree, and will add that same comment here if I need to touch this up again.

File: src/net_processing.cpp
Comment (mzumsande) [inline src/net_processing.cpp:2477]: nit: maybe mentioning the block hash and/or height in the unconditional logs here and below could be helpful so in case of a disk failure, users could reproduce the failure easier (e.g. with getblock RPC).

**Review by mzumsande — COMMENTED** [review]
Code Review ACK 75d27fefc7a04ebdda7be5724a014b6a896e7325

**Review by sedited — APPROVED** [review]
ACK 75d27fefc7a04ebdda7be5724a014b6a896e7325

**Comment by achow101** [comment]
ACK 75d27fefc7a04ebdda7be5724a014b6a896e7325

**Pull request merged**
